### PR TITLE
RFC - AI Avatar Animation Design

### PR DESCRIPTION
**Pull request by ctao456**
Add RFC for AI avatar animation project

[opea-project/GenAIComps#400](https://github.com/opea-project/GenAIComps/pull/400)
[opea-project/GenAIExamples#523](https://github.com/opea-project/GenAIExamples/pull/523)

### CODE REVIEW

File: community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
+## Title
```
Comment (louie-tsai) [inline community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md:1]:
```suggestion
##  A RAG-Powered Human-Like AI Avatar Audio Chatbot integrated with OPEA AudioQnA
```

File: community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+<!-- The short description of the feature you want to contribute -->
+Using OPEA to implement a RAG-Powered Human-Like AI Avatar Audio Chatbot (WIP)
```
Comment (louie-tsai) [inline community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md:3]:
```suggestion

```

File: community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+## Status
+<!-- Change the PR status to Under Review | Rejected | Accepted. -->
+v0.1 - ASMO Team sharing on Fri 6/28/2024  
+[GenAIComps pr #400](https://github.com/opea-project/GenAIComps/pull/400) (Under Review)
+[GenAIExamples pr #523](https://github.com/opea-project/GenAIExamples/pull/523) (Under Review)
+
+## Objective
+<!-- List what problem will this solve? What are the goals and non-goals of this RFC? -->
+* Introducing a new microservice, animation, that generates animated avatars from audio and image/video inputs; and a new megaservice, AvatarChatbot, that integrates the animation microservice with the existing AudioQnA service to build a human-like AI audio chatbot.
```
Comment (louie-tsai) [inline community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md:17]: possible to define how important to have AI avatar and what impact we might have if we don't have it? try to take some information from our competitor: https://nvidianews.nvidia.com/news/digital-humans-ace-generative-ai-microservices

File: community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+* This microservice leverages multimodal inputs (audio and image/video) and co-process them with a visual-audio fusion model, enriching OPEA with a new multimodal processing capability.
```
Comment (louie-tsai) [inline community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md:18]: Define a problem first. maybe missing multimodal support in OPEA

File: community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+* Enhancing OPEA's general availability of GenAI microservices for digital human, human-computer interaction, and conversational AI applications. 
+* Demonstrating the combination of multi-modal AI models and neural graphics rendering in a real-time avatar chatbot service
```
Comment (louie-tsai) [inline community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md:20]: if 2-3 pictures could explain the idea, no need to have that many.

File: community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+Related works include [Nvidia Audio2Face](https://docs.nvidia.com/ace/latest/modules/a2f-docs/index.html), [Lenovo Deepbrain AI Avatar](https://www.deepbrain.io/ai-avatars), [BitHuman](https://www.bithuman.io/), etc.
+
+## Design Proposal
+<!-- This is the heart of the document, used to elaborate the design philosophy and detail proposal. -->
+
+### Avatar Chatbot design
+[Proposal slides](https://intel.sharepoint.com/:p:/s/mlconsultingandsupport/EecmTa5Ca61Pm7ES6vlZtykBtgbJL_fuhipfu_yybIlBig?e=HDS0VF)
+
+<img src="assets/design.png" alt="Avatar Chatbot design" width="800"/>
```
Comment (louie-tsai) [inline community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md:58]: the Design doesn't match Flowchat.
Could we align both?  Moveover, what different messages between those two diagrams?

File: community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+Overall, this project adds to the OPEA platform a new microservice block that animates the chatbot appearance, and integrates it with the existing chatbot pipelines such as [ChatQnA](https://github.com/opea-project/GenAIExamples/tree/2e312f44edbcbf89bf00bc21d9e9c847405ecae8/ChatQnA), [AudioQnA](https://github.com/opea-project/GenAIExamples/tree/2e312f44edbcbf89bf00bc21d9e9c847405ecae8/AudioQnA), [SearchQnA](https://github.com/opea-project/GenAIExamples/tree/2e312f44edbcbf89bf00bc21d9e9c847405ecae8/SearchQnA), etc., to build new chatbot megaservices that can interact with users in a more human-like way.
+
+Related works include [Nvidia Audio2Face](https://docs.nvidia.com/ace/latest/modules/a2f-docs/index.html), [Lenovo Deepbrain AI Avatar](https://www.deepbrain.io/ai-avatars), [BitHuman](https://www.bithuman.io/), etc.
+
+## Design Proposal
+<!-- This is the heart of the document, used to elaborate the design philosophy and detail proposal. -->
+
+### Avatar Chatbot design
+[Proposal slides](https://intel.sharepoint.com/:p:/s/mlconsultingandsupport/EecmTa5Ca61Pm7ES6vlZtykBtgbJL_fuhipfu_yybIlBig?e=HDS0VF)
```
Comment (louie-tsai) [inline community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md:56]: don't put an internal link into public PR.

File: community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+User can build their own Docker image with `Dockerfile_hpu` and create a Docker container on Gaudi2 instance to run the animation microservice. They can then validate the service by sending a POST request to the server API, while providing audio and image/video inputs. The animation microservice will generate an animated avatar video response and save it to the specified output path.
+
+Support for alternative SoTA models such as [SadTalker](https://github.com/OpenTalker/SadTalker) and [LivePortrait](https://github.com/KwaiVGI/LivePortraitare) in progress.
+
+#### AvatarChatbot megaservice
+The AvatarChatbot megaservice is a new service that integrates the existing microservices that comprise AudioQnA service with the new animation microservice. The AudioQnA service is a pipeline that takes user audio input, converts it to text, generates an LLM response, and converts the response to audio output. The animation microservice is a new service that takes the audio response from the AudioQnA service, generates an animated avatar response, and sends it back to the user. The megaflow is as follows:  
+asr -> llm -> tts -> animation
```
Comment (louie-tsai) [inline community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md:102]: might be use mermaid flowchart here instead?

File: community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+This megaservice is a work in progress.
+
+#### Frontend UI
+The frontend UI is Gradio. User is prompted to upload either an image or a video as the avatar source. The user also asks his question verbally through the microphone. The AvatarChatbot backend processes the audio input and generates the response in the form of an animated avatar answering in its unique voice. The response is displayed on Gradio UI.
```
Comment (louie-tsai) [inline community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md:107]: We might also give some instructions how to use UI. ex: click audio section to record the your question.

File: community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+<div style="display: flex; justify-content: space-between;">
+  <img src="assets/ui_1.png" alt="alt text" style="width: 49%;"/>
+  <img src="assets/ui_2.png" alt="alt text" style="width: 49%;"/>
+</div>
+
+### Real-time demo
+[AI Avatar Chatbot Demo on Intel® Gaudi® 2, image input](https://intel.sharepoint.com/:v:/s/mlconsultingandsupport/EZa7vjON10ZCpMvE7U-SPMwBRXbVHqe1Ybsa-fmnXayNUA?e=f6FPsl)
```
Comment (louie-tsai) [inline community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md:114]: don't put the internal links here. could we have motion jpeg here?

**Review by louie-tsai — COMMENTED** [review]
Great works. put some comments in PR.

File: community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md (hunk 6 above)
Comment (ctao456) [reply community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md:60]: The difference is that the Design includes RAG feature, that will be released in v0.2.  
The Flowchart is based on v0.1.
Explained accordingly in the revision.

File: community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md (hunk 3 above)
Comment (ctao456) [reply community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md]: Revised accordingly

File: community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md (hunk 8 above)
Comment (ctao456) [reply community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md:104]: Included a more accurate mermaid flowchart for the new `AvatarChatbot` megaservice.

File: community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md (hunk 10 above)
Comment (ctao456) [reply community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md]: Switched to mpg files and rendered in the markdown file

File: community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md (hunk 9 above)
Comment (ctao456) [reply community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md]: Revised accordingly

File: community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md [hunk 11]
```diff
@@ -0,0 +1,222 @@
+##  A RAG-Powered Human-Like AI Avatar Audio Chatbot integrated with OPEA AudioQnA
```
Comment (dbkinder) [inline community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md:1]: Title should be the RFC name
```suggestion
# 24-08-02-OPEA-AIAvatarChatbot

A RAG-Powered Human-Like AI Avatar Audio Chatbot integrated with OPEA AudioQnA
```

**Review by dbkinder — CHANGES_REQUESTED** [review]
Just a title suggested fix

**Review by dbkinder — APPROVED** [review]
LGTM, thanks.

**Review by dbkinder — COMMENTED** [review]
Looks like you didn't do a "git commit -s" so your PR is missing the required Signed-off-by: line in the commit message :(

**Comment by ctao456** [comment]
> Looks like you didn't do a "git commit -s" so your PR is missing the required Signed-off-by: line in the commit message :(

Thank you. That was true. I was using the web interface to commit. Should never use it. I fixed it by rebasing to HEAD~2 and commited with sign-off, then force pushing.

**Review by mkbhanda — APPROVED** [review]
LGTM

File: community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+<div style="display: flex; justify-content: space-between;">
+  <img src="assets/ui_latest_1.png" alt="alt text" style="width: 33%;"/>
+  <img src="assets/ui_latest_2.png" alt="alt text" style="width: 33%;"/>
+  <img src="assets/ui_latest_3.png" alt="alt text" style="width: 33%;"/>
+</div>
+
+### Real-time demo
+AI Avatar Chatbot Demo on Intel® Gaudi® 2, image input (top) and video input (down)
+<!-- [AI Avatar Chatbot Demo on Intel® Gaudi® 2, image input](https://intel.sharepoint.com/:v:/s/mlconsultingandsupport/EZa7vjON10ZCpMvE7U-SPMwBRXbVHqe1Ybsa-fmnXayNUA?e=f6FPsl)   
+[AI Avatar Chatbot Demo on Intel® Gaudi® 2, video input](https://intel.sharepoint.com/:v:/s/mlconsultingandsupport/ESMIcBseFTdIuqkoB7TZy6ABfwR9CkfV49TvTa1X_Jihkg?e=zMH7O7) -->
```
Comment (ZePan110) [inline community/rfcs/24-08-02-OPEA-AIAvatarChatbot.md:193]: @ctao456  We shouldn't have such an URL in our documentation (it's not accessible to the community). If that video is important, it should be put in a public location.

**Pull request merged**